src/daft-local-execution/src/streaming_sink/vllm.rs

```diff
@@ -0,0 +1,361 @@
+                    DaftError::type_error(format!(
+                        "Expected input to `prompt` to be string, got {}",
+                        prompt_series.data_type()
+                    ))
+                })?
+                .as_arrow()
+                .values_iter()
+                .collect::<Vec<_>>();
+
+            state.executor.submit(prompts, rb)?;
+        }
+
+        let null_rows = null_rows
+            .into_iter()
+            .map(|rb| {
+                rb.append_column(
+                    self.schema.clone(),
+                    Series::full_null(&self.output_column_name, &DataType::Utf8, rb.len()),
+                )
+                .unwrap()
+            })
+            .collect::<Vec<_>>();
+
+        Ok(null_rows)
+    }
+
+    fn poll_tasks(&self, state: &VLLMState) -> DaftResult<Option<RecordBatch>> {
+        Ok(state.executor.poll()?.map(|(outputs, rows)| {
+            let output_series =
+                Utf8Array::from((self.output_column_name.as_ref(), outputs.as_slice()))
+                    .into_series();
+
+            rows.append_column(self.schema.clone(), output_series)
+                .unwrap()
+        }))
+    }
+
+    fn combine_output_with_null_rows(
+        &self,
+        output_batch: Option<RecordBatch>,
+        null_rows: Vec<RecordBatch>,
+    ) -> Option<Arc<MicroPartition>> {
+        if let Some(output_batch) = output_batch {
+            let mut output_batches = null_rows;
+            output_batches.push(output_batch);
+
+            Some(Arc::new(MicroPartition::new_loaded(
+                self.schema.clone(),
+                Arc::new(output_batches),
+                None,
+            )))
+        } else if null_rows.is_empty() {
+            None
+        } else {
+            Some(Arc::new(MicroPartition::new_loaded(
+                self.schema.clone(),
+                Arc::new(null_rows),
+                None,
+            )))
+        }
+    }
+
+    fn try_add_to_buffer(&self, state: &mut VLLMState, input: Arc<MicroPartition>) -> bool {
+        if state.buffer_size < self.expr.inner().max_buffer_size {
+            state.buffer_size += input.len();
+            state.buffer.push_back(input);
+
+            true
+        } else {
+            false
+        }
+    }
+}
+
+impl StreamingSink for VLLMSink {
+    type State = VLLMState;
+
+    fn execute(
+        &self,
+        input: Arc<MicroPartition>,
+        mut state: Self::State,
+        spawner: &ExecutionTaskSpawner,
+    ) -> StreamingSinkExecuteResult<Self> {
+        let this = self.clone();
+
+        spawner
+            .spawn(
+                async move {
+                    let input_added = this.try_add_to_buffer(&mut state, input.clone());
+
+                    let null_rows = this.submit_tasks(&mut state)?;
+
+                    let output_batch = this.poll_tasks(&state)?;
+                    let output = this.combine_output_with_null_rows(output_batch, null_rows);
```

Comment:
Comments explaining the flow or docstrings on the helper functions would be helpful

Reply:
same as aboe